File: enve/Services/Player/PlayerBookmarkService.swift
import Foundation

public class PlayerBookmarkService {
    private let storageService: StorageService

    public init(storageService: StorageService = StorageService()) {
        self.storageService = storageService
    }

    public func loadBookmarks(bookId: String) -> [Bookmark] {
        return ReaderArtifactsStore.shared.loadBookmarks(bookId: bookId)
    }

    public func addBookmark(
        bookId: String,
        position: TimeInterval,
        locator: String?,
        title: String?,
        note: String?,
        mediaType: AppMediaType,
        chapterTitle: String? = nil,
        remoteID: Int? = nil,
        isRemotePlaceholder: Bool = false
    ) -> Bookmark {
        let autoTitle: String
        if let title = title, !title.isEmpty {
            autoTitle = title
        } else if mediaType == .ebook {
            autoTitle = chapterTitle ?? "Bookmark at \(Int(position * 100))%"
        } else if let chapter = chapterTitle, !chapter.isEmpty {
            autoTitle = "Bookmark - \(chapter)"
        } else {
            autoTitle = "Bookmark at \(formatTime(position))"
        }
        let bookmark = Bookmark(
            bookId: bookId,
            position: position,
            title: autoTitle,
            note: note,
            timestamp: Date(),
            locator: locator,
            mediaType: mediaType,
            chapterTitle: chapterTitle,
            remoteID: remoteID,
            isRemotePlaceholder: isRemotePlaceholder
        )
        var bookmarks = ReaderArtifactsStore.shared.loadBookmarks(bookId: bookId)
        bookmarks.append(bookmark)
        ReaderArtifactsStore.shared.saveBookmarks(bookId: bookId, bookmarks: bookmarks)
        return bookmark
    }

    public func addBookmark(bookId: String, position: TimeInterval, title: String?, note: String?) -> Bookmark {
        return addBookmark(bookId: bookId, position: position, locator: nil, title: title, note: note, mediaType: .audiobook)
    }

    public func deleteBookmark(_ bookmark: Bookmark) {
        var bookmarks = ReaderArtifactsStore.shared.loadBookmarks(bookId: bookmark.bookId)
        bookmarks.removeAll { $0.id == bookmark.id }
        ReaderArtifactsStore.shared.saveBookmarks(bookId: bookmark.bookId, bookmarks: bookmarks)
    }

    public func updateBookmark(_ bookmark: Bookmark) {
        var bookmarks = ReaderArtifactsStore.shared.loadBookmarks(bookId: bookmark.bookId)
        if let index = bookmarks.firstIndex(where: { $0.id == bookmark.id }) {
            bookmarks[index] = bookmark
        }
        ReaderArtifactsStore.shared.saveBookmarks(bookId: bookmark.bookId, bookmarks: bookmarks)
    }

    public func replaceBookmarks(bookId: String, bookmarks: [Bookmark]) {
        ReaderArtifactsStore.shared.saveBookmarks(bookId: bookId, bookmarks: bookmarks)
    }

    private func formatTime(_ seconds: TimeInterval) -> String {
        let hours = Int(seconds) / 3600
        let minutes = (Int(seconds) % 3600) / 60
        let remainingSeconds = Int(seconds) % 60

        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, remainingSeconds)
        } else {
            return String(format: "%d:%02d", minutes, remainingSeconds)
        }
    }
}

File: android/engine/src/main/java/com/enve/app/playback/PlayerBookmarkService.kt
package com.enve.app.playback

import android.content.Context
import com.enve.core.data.model.AppMediaType
import com.enve.core.data.model.AudiobookBookmark
import com.enve.core.data.model.Book
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PlayerBookmarkService @Inject constructor(
    @ApplicationContext private val context: Context,
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    @Serializable
    private data class BookmarkPayload(val bookmarks: List<AudiobookBookmark> = emptyList())

    suspend fun loadBookmarks(book: Book): List<AudiobookBookmark> = readBookmarks(book.id)

    suspend fun addBookmark(
        book: Book,
        position: Long,
        title: String?,
        note: String?,
        chapterTitle: String?,
    ): AudiobookBookmark {
        val bookmark = AudiobookBookmark(
            bookId = book.id,
            position = position.coerceAtLeast(0),
            title = buildTitle(
                requestedTitle = title,
                mediaType = book.mediaType,
                position = position,
                chapterTitle = chapterTitle,
            ),
            note = note?.takeIf { it.isNotBlank() },
            mediaType = book.mediaType,
            chapterTitle = chapterTitle?.takeIf { it.isNotBlank() },
        )
        upsertLocal(book.id, bookmark)
        return bookmark
    }

    suspend fun updateBookmark(book: Book, bookmark: AudiobookBookmark, title: String, note: String?): AudiobookBookmark {
        val updated = bookmark.copy(
            title = title.ifBlank { bookmark.title },
            note = note?.takeIf { it.isNotBlank() },
        )
        upsertLocal(book.id, updated)
        return updated
    }

    suspend fun deleteBookmark(book: Book, bookmark: AudiobookBookmark) {
        val remaining = readBookmarks(book.id).filterNot { it.id == bookmark.id }
        writeBookmarks(book.id, remaining)
    }

    fun buildTitle(
        requestedTitle: String?,
        mediaType: AppMediaType,
        position: Long,
        chapterTitle: String?,
    ): String {
        if (!requestedTitle.isNullOrBlank()) return requestedTitle.trim()
        if (mediaType == AppMediaType.EBOOK) {
            return chapterTitle?.takeIf { it.isNotBlank() } ?: "Bookmark at ${(position * 100).toInt()}%"
        }
        if (!chapterTitle.isNullOrBlank()) return "Bookmark: $chapterTitle"
        return "Bookmark at ${AudiobookBookmark.formatTime(position)}"
    }

    private suspend fun upsertLocal(bookId: String, bookmark: AudiobookBookmark) {
        val updated = readBookmarks(bookId).toMutableList()
        val index = updated.indexOfFirst { it.id == bookmark.id }
        if (index >= 0) updated[index] = bookmark else updated.add(bookmark)
        writeBookmarks(bookId, updated)
    }

    private suspend fun readBookmarks(bookId: String): List<AudiobookBookmark> = withContext(Dispatchers.IO) {
        runCatching {
            val file = bookmarkFile(bookId)
            if (!file.exists() || file.length() == 0L) return@runCatching emptyList()
            json.decodeFromString<BookmarkPayload>(file.readText()).bookmarks
        }.getOrDefault(emptyList())
    }

    private suspend fun writeBookmarks(bookId: String, bookmarks: List<AudiobookBookmark>) = withContext(Dispatchers.IO) {
        val ordered = bookmarks.sortedBy { it.timestamp }
        val target = bookmarkFile(bookId)
        target.parentFile?.mkdirs()
        val tmp = File(target.parentFile, "${target.name}.tmp")
        tmp.writeText(json.encodeToString(BookmarkPayload(ordered)))
        if (!tmp.renameTo(target)) {
            target.writeText(tmp.readText())
            tmp.delete()
        }
    }

    private fun bookmarkFile(bookId: String): File {
        val safeBookId = bookId.replace(Regex("[^A-Za-z0-9._-]"), "_")
        return File(context.filesDir, "audiobook-bookmarks/$safeBookId.json")
    }
}
